File: 34-n-queens/dfs.cc
class Solution {
public:
    /**
     * @param n: The number of queens.
     * @return: The total number of distinct solutions.
     */
    int totalNQueens(int n) {
        // write your code here
        if (n <= 0) {
            return 0;
        }
        int result = 0;
        vector<int> col;
        search(result, col, n);
        return result;
    }
    void search(int& result, vector<int>& col, int n) {
        if (col.size() == n) {
            result++;
        }
        for (int i = 0; i < n; i++) {
            if (!isValid(col, i)) {
                continue;
            }
            col.push_back(i);
            search(result, col, n);
            col.pop_back();
        }
    }
    bool isValid(vector<int>& col, int colomn) {
        int row = col.size();
        for (int i = 0; i < col.size(); i++) {
            if (col[i] == colomn) {
                return false;
            }
            if (row - i == colomn - col[i]) {
                return false;
            }
            if (row - i == col[i] - colomn) {
                return false;
            }
        }
        return true;
    }
};

File: 11-search-range-in-binary-search-tree/dfs.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: param root: The root of the binary search tree
     * @param k1: An integer
     * @param k2: An integer
     * @return: return: Return all keys that k1<=key<=k2 in ascending order
     */
    private List<Integer> result = new ArrayList<Integer>();
    public List<Integer> searchRange(TreeNode root, int k1, int k2) {
        // write your code here
        if (root == null || k1 > k2) {
            return result;
        }
        findElement(root, k1, k2);
        return result;
    }
    void findElement(TreeNode root, int k1, int k2) {
        if (root == null) {
            return;
        }
        if (root.val < k1) {
            findElement(root.right, k1, k2);
        } else if (root.val > k2) {
            findElement(root.left, k1, k2);
        } else {
            result.add(root.val);
            findElement(root.left, k1, k2);
            findElement(root.right, k1, k2);
        }
    }
}
